pre_commit_all_the_way_down.py

```diff
@@ -0,0 +1,126 @@
+import argparse
+import re
+import subprocess
+import sys
+from pathlib import Path
+from re import Match
+from tempfile import TemporaryDirectory
+from textwrap import dedent, indent
+from typing import Optional, Sequence
+
+BLOCK_TYPES = "(code|code-block|sourcecode|ipython)"
+PY_LANGS = "(python|py|sage|python3|py3|numpy)"
+DOCTEST_TYPES = "(testsetup|testcleanup|testcode)"
+RST_RE = re.compile(
+    rf"(?P<before>"
+    rf"^(?P<indent> *)\.\. ("
+    rf"jupyter-execute::|"
+    rf"python-script::|"
+    rf"{BLOCK_TYPES}:: {PY_LANGS}|"
+    rf"{DOCTEST_TYPES}::.*"
+    rf")\n"
+    rf"((?P=indent) +:.*\n)*"
+    rf"\n*"
+    rf")"
+    rf"(?P<code>(^((?P=indent) +.*)?\n)+)",
+    re.MULTILINE,
+)
+INDENT_RE = re.compile("^ +(?=[^ ])", re.MULTILINE)
+TRAILING_NL_RE = re.compile(r"\n+\Z", re.MULTILINE)
+
+
+def fix_block(block: str, whitelist: Sequence[Optional[str]]):
+    """
+    Fix a code block.
+
+    Parameters
+    ----------
+    block : str
+        The block to fix
+
+    Notes
+    -----
+    The code block should have no leading indentation and be valid
+    Python.
+    """
+
+    def _pre_commit_helper(fname: str, hook_id: Optional[str]):
+        args = ["pre-commit", "run"]
+        if hook_id is not None:
+            args.append(hook_id)
+        args.extend(("--files", fname))
+        try:
+            subprocess.run(args, check=True, capture_output=True)
+        except subprocess.CalledProcessError as e:
+            print(e.stderr.decode(), file=sys.stderr)
+            print(e.stdout.decode(), file=sys.stdout)
```

Comment:
I mean I get the symmetry here, maybe I'm just allergic to patterns that look like duplicated code :)
```suggestion
            print(e.stdout.decode())
```